### PR TITLE
node 4x retry

### PR DESCRIPTION
**Pull request by erulabs**
Trying via https://gist.github.com/thlorenz/7e9d8ad15566c99fd116

### CODE REVIEW

**Comment by erulabs** [comment]
@sonnyp Maybe you can take a peek at this one? Should be a lot closer.

I think all the Nan2 stuff is working - but the core tests are failing :/

**Comment by erulabs** [comment]
Hey @sonnyp 

  Well, I'm close. Very close. Can't figure how why it's not throwing exceptions for fallback JS profiles properly...

  I've got to stop working on this now (ran out of time). HUUUGE thanks goes out to anyone who spends some time on this for me :)

Thanks!

File: index.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ var UIDNA_ALLOW_UNASSIGNED = 1
 var UIDNA_USE_STD3_RULES = 2
 
 try {
-    var bindings = require('bindings')('node_stringprep.node')
+    var bindings = require('./node_modules/bindings/bindings.js')('node_stringprep.node')
```
Comment (rvagg) [inline index.js]: whoa, no, why was this changed?

File: node-stringprep.cc [hunk 2]
```diff
@@ -15,17 +15,17 @@ class UnknownProfileException : public std::exception {
...
-    NanAssignPersistent(stringprep_constructor, t);
+    Nan::HandleScope scope;
+    Local<FunctionTemplate> t = Nan::New<FunctionTemplate>(New);
+    //stringprep_constructor.Reset( t);
     t->InstanceTemplate()->SetInternalFieldCount(1);
-    NODE_SET_PROTOTYPE_METHOD(t, "prepare", Prepare);
+    Nan::SetPrototypeMethod(t, "prepare", Prepare);
 
-    target->Set(NanNew<String>("StringPrep"), t->GetFunction());
+    target->Set(Nan::New<String>("StringPrep").ToLocalChecked(), t->GetFunction());
```
Comment (rvagg) [inline node-stringprep.cc]: use `Nan::Set()` for this

File: node-stringprep.cc [hunk 3]
```diff
@@ -43,11 +43,11 @@ class StringPrep : public ObjectWrap {
 
   static NAN_METHOD(New)
   {
-    NanScope();
+    Nan::HandleScope scope;
```
Comment (rvagg) [inline node-stringprep.cc]: remove

File: node-stringprep.cc [hunk 4]
```diff
@@ -43,11 +43,11 @@ class StringPrep : public ObjectWrap {
...
 
-    if (args.Length() >= 1 && args[0]->IsString())
+    if (info.Length() >= 1 && info[0]->IsString())
       {
-        String::Utf8Value arg0(args[0]->ToString());
+        String::Utf8Value arg0(info[0]->ToString());
```
Comment (rvagg) [inline node-stringprep.cc]: `Nan::Utf8Value`

File: node-stringprep.cc [hunk 5]
```diff
@@ -55,27 +55,27 @@ class StringPrep : public ObjectWrap {
...
+            return;
           }
 
         StringPrep *self = new StringPrep(profileType);
         if (self->good())
           {
-            self->Wrap(args.This());
-            NanReturnValue(args.This());
+            self->Wrap(info.This());
+            info.GetReturnValue().Set(info.This());
```
Comment (rvagg) [inline node-stringprep.cc]: `return` - not strictly necessary but you're returning in other branches so be safe and communicate intention to future devs

File: node-stringprep.cc [hunk 6]
```diff
@@ -97,17 +97,17 @@ class StringPrep : public ObjectWrap {
 
   static NAN_METHOD(Prepare)
   {
-    NanScope();
+    Nan::HandleScope scope;
```
Comment (rvagg) [inline node-stringprep.cc]: remove

File: node-stringprep.cc [hunk 7]
```diff
@@ -97,17 +97,17 @@ class StringPrep : public ObjectWrap {
...
 
-    if (args.Length() >= 1 && args[0]->IsString())
+    if (info.Length() >= 1 && info[0]->IsString())
       {
-        StringPrep *self = ObjectWrap::Unwrap<StringPrep>(args.This());
-        String::Value arg0(args[0]->ToString());
-        NanReturnValue(self->prepare(arg0));
+        StringPrep *self = Nan::ObjectWrap::Unwrap<StringPrep>(info.This());
+        String::Value arg0(info[0]->ToString());
+        info.GetReturnValue().Set(self->prepare(arg0));
```
Comment (rvagg) [inline node-stringprep.cc]: return

**Comment by rvagg** [comment]
You _should_ be able to replace all references to `Handle` with `Local` for maximum compatibility

**Comment by rvagg** [comment]
` Handle<Value> prepare(String::Value &str)` should have a `Nan::HandleScope scope;` otherwise it'll leak data (looks like it has been till now anyway) and in this case it should be a `Nan::EscapableHandleScope scope` because you're returning a value from it.

**Comment by rvagg** [comment]
https://github.com/nodejs/nan/blob/master/doc/scopes.md#nanescapablehandlescope - you also need to use `scope.Escape(obj)`

File: node-stringprep.cc [hunk 8]
```diff
@@ -131,20 +131,19 @@ class StringPrep : public ObjectWrap {
...
           }
-        else if (!good())
+        else if (!U_SUCCESS(error))
           {
             // other error, just bail out
             delete[] dest;
-            NanThrowError(errorName());
-            return NanUndefined();
+            Nan::ThrowError(u_errorName(error));
+            return Nan::Undefined();
```
Comment (rvagg) [inline node-stringprep.cc]: this is the `scope.Escape(Nan::Undefined())`

File: node-stringprep.cc [hunk 9]
```diff
@@ -131,20 +131,19 @@ class StringPrep : public ObjectWrap {
...
           }
         else
           destLen = w;
       }
 
-    Local<String> result = NanNew<String>(dest, destLen);
     delete[] dest;
-    return result;
+    return Nan::New<String>(dest, destLen).ToLocalChecked();
```
Comment (rvagg) [inline node-stringprep.cc]: and `scope.Escape(Nan::New<String>(dest, destLen).ToLocalChecked())`

File: node-stringprep.cc [hunk 10]
```diff
@@ -192,12 +191,12 @@ class StringPrep : public ObjectWrap {
 
 NAN_METHOD(ToUnicode)
 {
-  NanScope();
+  Nan::HandleScope scope;
```
Comment (rvagg) [inline node-stringprep.cc]: remove

File: node-stringprep.cc [hunk 11]
```diff
@@ -222,31 +221,31 @@ NAN_METHOD(ToUnicode)
...
           }
         else
           destLen = w;
       }
 
-    Local<String> result = NanNew<String>(dest, destLen);
+    Local<String> result = Nan::New<String>(dest, destLen).ToLocalChecked();
     delete[] dest;
-    NanReturnValue(result);
+    info.GetReturnValue().Set(result);
```
Comment (rvagg) [inline node-stringprep.cc]: return

File: node-stringprep.cc [hunk 12]
```diff
@@ -222,31 +221,31 @@ NAN_METHOD(ToUnicode)
...
-    NanReturnUndefined();
+    Nan::ThrowTypeError("Bad argument.");
+    return;
   }
 }
 
 NAN_METHOD(ToASCII)
 {
-  NanScope();
+  Nan::HandleScope scope;
```
Comment (rvagg) [inline node-stringprep.cc]: remove

File: node-stringprep.cc [hunk 13]
```diff
@@ -270,28 +269,29 @@ NAN_METHOD(ToASCII)
...
-        NanReturnUndefined();
+        Nan::ThrowError(u_errorName(error));
+        return;
       }
 
-    Local<String> result = NanNew<String>(dest, destLen);
+    Local<String> result = Nan::New<String>(dest, destLen).ToLocalChecked();
     delete[] dest;
-    NanReturnValue(result);
+    info.GetReturnValue().Set(result);
```
Comment (rvagg) [inline node-stringprep.cc]: return

File: node-stringprep.cc [hunk 14]
```diff
@@ -270,28 +269,29 @@ NAN_METHOD(ToASCII)
...
 /*** Initialization ***/
 
-extern "C" void init(Handle<Object> target)
-{
-  NanScope();
-  StringPrep::Initialize(target);
-  NODE_SET_METHOD(target, "toUnicode", ToUnicode);
-  NODE_SET_METHOD(target, "toASCII", ToASCII);
+extern "C" {
+  static void init (Handle<Object> target)
```
Comment (rvagg) [inline node-stringprep.cc]: `Handle` will end up being a problem here (it's being deprecated so you'll end up with errors in a version of Node or two), opt for declaring this as `NAN_MODULE_INIT(init)` anyway and you'll be safe

**Comment by rvagg** [comment]
I did a write-up @ https://nodesource.com/blog/cpp-addons-for-nodejs-v4 about the upgrade process that's a bit more verbose than other resources currently available

**Comment by sonnyp** [comment]
closing this in fabor of https://github.com/node-xmpp/node-stringprep/pull/72
